benchadapt/python/benchadapt/result.py

```diff
@@ -210,6 +214,48 @@ def to_publishable_dict(self):
         return res_dict
 
 
+def validate_or_remove_github_commit_key(res_dict: Dict, strict=False):
+    """
+    Mutate BenchmarkResult dictionary in-place to make its `github` key
+    property be compliant with the Conbench HTTP API.
+
+    Not providing the `github` key in result dictionary tells Conbench that
+    this result is commit context-less (recording it in Conbench might be
+    useful for debugging and testing purposes, but generally we should make
+    clear that this implies missing out on critical features/purpose).
+    """
+
+    def _warn_or_raise():
+        msg = (
+            "Entity not publishable! `github.repository` and `github.commit` "
+            "must be populated. You may pass github metadata via "
+            "CONBENCH_PROJECT_REPOSITORY, CONBENCH_PROJECT_COMMIT, "
+            "and CONBENCH_PR_NUMBER environment variables. "
+            f"\ngithub: {res_dict['github']}"
+        )
+        if strict:
+            raise ValueError(msg)
+
+        warnings.warn(msg)
+
+    # For now, the decision is to send the result out, signaling to the
+    # Conbench API that this result has no repo/commit context. Maybe we should
+    # have the client tooling error out instead?
+    if "github" not in res_dict:
+        # API-wise, that's a valid state, signaling "no commit information
+        # present"
+        _warn_or_raise()
+        return
+
+    for checkkey in ("repository", "commit"):
+        if checkkey not in res_dict["github"]:
+            _warn_or_raise()
```

Comment:
Also, I think we should still support the first case even for runs. So this should not error out, like it does right now:
```py
BenchmarkRun().to_publishable_dict()
# ValueError: Entity not publishable! `github.repository` and `github.commit` must be populated. You may pass github metadata via CONBENCH_PROJECT_REPOSITORY, CONBENCH_PROJECT_COMMIT, and CONBENCH_PR_NUMBER environment variables.
# github: {}
```
We'd like to be able to post a no-context commit.

Reply:
For a no-context commit, do we want to pass `"github": {}` or just not populate it? In parallel situations (e.g. `error`, `stats`) we pop off empty dicts.

Also, this interface for opting-in to no-context commits (not specifying `github` when no env vars are set) is kinda crappy. An explicit `github={}` (even if we want to pop it off) would be nicer.

And frankly I think usually `repo` should still be set for most of these cases even if `commit`, `pr_number`, and `branch` are empty, but that's maybe bigger than the scope of this PR.

Reply:
> For a no-context commit, do we want to pass "github": {} or just not populate it? 

I take the current state of what the HTTP wants (how it's specified) for granted for the moment. We can always review/refine/change that part, but this here is basically adjusting client tooling to the latest HTTP API change. 

Everything becomes easier when the following is true: if a `github` key is provided, then _it is a full specification of a commit (long hash) in a repository in github (specified by URL)_; no exceptions.

Madness happens given all the various possible permutations between
- no dict allowed
- empty dict allowed
- if dict specified:
    - only commit allowed
    - only repo allowed
    - if repo specified:
        - url allowed
        - org/repo notation allowed

(and more) -- this is a _huge_ permutation space, and it did leak deep into our business logic. We didn't try to normalize this upon receiving the HTTP request, we did put most of this complexity even into the database. 

I just want to reduce the complexity for sanity.

Reply:
> And frankly I think usually repo should still be set for most of these cases even if commit, pr_number, and branch are empty, but that's maybe bigger than the scope of this PR.

This is probably correct thinking and heavily relates to https://github.com/conbench/conbench/issues/1165, right?

Reply:
> I don't think the warning should say the result isn't publishable. It is publishable now! 

thanks for confirming, yes. I didn't want to go to deep here with adjusting messages, but I think now I do.

Reply:
Thanks for the code examples, too. Thinking about this a little more.

When having env vars set and when
- not providing an argument for `github` then the outcome should be populated from env.
- providing `github=None` that's the explicit way to say "no commit information"

Distinguishing both cases can probably be easily achieved with https://docs.python.org/3/library/dataclasses.html#post-init-processing (if that's needed at all).

Reply:
I tried to avoid `__post_init__`s with these because you can end up with different behavior if you set something on instantiation vs after, which does happen with these, e.g. https://github.com/conbench/conbench/blob/e23b5fee689034c1cd99f92061cce92d604f59a1/benchadapt/python/benchadapt/adapters/_adapter.py#L153-L161

> * providing `github=None` that's the explicit way to say "no commit information"

That's shorthand for using the autodetection (which in retrospect was a regrettable decision; maybe we can unwind?). I think you need `github={}` to really specify "no repo/commit info".

Reply:
As far as I'm aware, no one is using `github=None` for autodetection (scraping git locally) today. It's still a feature I want to keep around but maybe we can move that to a different opt-in interface and use `github=None` for "no commit information"? (though if we do want to do something around #1165 in the future, we'll need to think about that interface too)

Reply:
Great discussion!

> That's shorthand for using the autodetection

Oh, and with that you mean the "inspect git repo in current working directory", I see. 

I see that is documented, the docstring says:

> Advanced: if you have a locally cloned repo, you may explicitly supply ``None``

Just for the record, dataclass is able to distinguish between "no value passed" and "`None` passed".

And we have code that handles the `None` case already [here](https://github.com/conbench/conbench/blob/e23b5fee689034c1cd99f92061cce92d604f59a1/benchadapt/python/benchadapt/result.py#L159): 

```
        if value is None:
            value = _machine_info.detect_commit_info_from_local_git()
```
OK, about those two statements:

> maybe we can unwind?).
> no one is using github=None for autodetection (scraping git locally) today

Let's add a new value for that, e.g. "inspect_git_in_cwd" (I would like the explicit nature of that). And I mean, we could also remove that functionality because that's very magic anyway and nobody uses it, but for now I would totally feel comfortable just to change how to "activate" this.

Then we can repurpose `None` as described above and also do not need `__post_init__`

Reply:
I want to get unstuck here, trying to put this into some commits and update the branch, so I can call it a day :). :sun_with_face:

Reply:
Lovely!